app/src/main/java/org/thoughtcrime/securesms/jobs/SubscriptionReceiptRequestResponseJob.java

```diff
@@ -0,0 +1,275 @@
+
+      ReceiptSerial             receiptSerial = new ReceiptSerial(randomBytes);
+      ClientZkReceiptOperations operations    = ApplicationDependencies.getClientZkReceiptOperations();
+
+      requestContext = operations.createReceiptCredentialRequestContext(secureRandom, receiptSerial);
+    }
+
+    ServiceResponse<ReceiptCredentialResponse> response = ApplicationDependencies.getDonationsService()
+                                                                                 .submitReceiptCredentialRequest(subscriberId, requestContext.getRequest())
+                                                                                 .blockingGet();
+
+    if (response.getApplicationError().isPresent()) {
+      handleApplicationError(response);
+    } else if (response.getResult().isPresent()) {
+      ReceiptCredential receiptCredential = getReceiptCredential(response.getResult().get());
+
+      if (!isCredentialValid(subscription, receiptCredential)) {
+        throw new IOException("Could not validate receipt credential");
+      }
+
+      ReceiptCredentialPresentation receiptCredentialPresentation = getReceiptCredentialPresentation(receiptCredential);
+      setOutputData(new Data.Builder().putBlobAsString(DonationReceiptRedemptionJob.INPUT_RECEIPT_CREDENTIAL_PRESENTATION,
+                                                       receiptCredentialPresentation.serialize())
+                                      .build());
+    } else {
+      Log.w(TAG, "Encountered a retryable exception: " + response.getStatus(), response.getExecutionError().orNull(), true);
+      throw new RetryableException();
+    }
+  }
+
+  private @Nullable ActiveSubscription.Subscription getLatestSubscriptionInformation() throws Exception {
+    ServiceResponse<ActiveSubscription> activeSubscription = ApplicationDependencies.getDonationsService()
+                                                                                    .getSubscription(subscriberId)
+                                                                                    .blockingGet();
+
+    if (activeSubscription.getResult().isPresent()) {
+      return activeSubscription.getResult().get().getActiveSubscription();
+    } else if (activeSubscription.getApplicationError().isPresent()) {
+      Log.w(TAG, "Unrecoverable error getting the user's current subscription. Failing.", activeSubscription.getApplicationError().get(), true);
+      throw new IOException(activeSubscription.getApplicationError().get());
+    } else {
+      throw new RetryableException();
+    }
+  }
+
+  private ReceiptCredentialPresentation getReceiptCredentialPresentation(@NonNull ReceiptCredential receiptCredential) throws RetryableException {
+    ClientZkReceiptOperations operations = ApplicationDependencies.getClientZkReceiptOperations();
+
+    try {
+      return operations.createReceiptCredentialPresentation(receiptCredential);
+    } catch (VerificationFailedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)